### ISSUE TITLE
tm_basemap(NULL) doesn't disable basemaps

### ISSUE THREAD

**Issue by barryrowlingson** [r-tmap/tmap#972]
Help page for `tm_basemap` says:

 Use   ‘NULL’ in ‘tm_basemap()’ to disable basemaps.

But in both plot and view mode I still get an ESRI base map. I would expect to not see a basemap. Obvious fix is
to not have the tm_basemap() call if you don't want a basemap, but that's harder to do if the basemap is driven
by a variable, eg `tm_basemap(whichone) + tm_shape(...)  + ...`

Ideally I think `tm_basemap` should be able to specify with arguments no basemap, the default basemap, or a
specific basemap. To preserve backward compatibility I suggest:

 * `tm_basemap()` : adds default basemap
 * `tm_basemap(NULL)` : adds default basemap, change documentation to clarify
 * `tm_basemap(NA)` : no basemap, think of it as a "missing value".
 * `tm_basemap(xyz)` : add basemap for provider xyz, xyz may also be NULL or NA to get behaviour as above.

Having no `tm_basemap` in a map would also remain the current behaviour which I think is default basemap in view mode, no basemap in plot mode. But I don't know if this minor inconsistency is worth preserving going forward...

```
> packageVersion("tmap")
[1] ‘3.99.9003’
```

**Comment by mtennekes** [comment]
Thx for bringing this up!

FYI: the default options for tmap are mode-specific. The mode-specific options are accessible via `tmap_options("modes")`. The function `tmap_options_mode()` returns all the options for a specific mode. It's not possible to easily change the default options for a mode. I'm still looking to make this more user-friendly...

There is an option called `basemap.show`. When `TRUE` (view mode only) it adds `tm_basemap()`. However, it does not remove added basemaps. So `tm_shape(World) + tm_polygons() + tm_basemap() + tm_options(basemap.show = FALSE)` will still show a basemap (actually a choice of three since `tmap_options("basemap.server")` contains three values.

Apart from the current bug, should we use NULL or NA? One should be used for "not specified, just use the default" and the other for "non-existent". I would propose:

`NULL`: non-existent
`NA`: non-specified (so: use default values)

This is also the behaviour in tmap3.

**This question is also relevant for visual variables:**

```r
tm_shape(World) +
	tm_polygons(fill = NA) 
```

could/should use the default fill color for polygons (default is grey, but theme dependent)


```
tm_shape(World) +
	tm_polygons(fill = NULL) 
```

should not fill the polygons

Does this make sense @barryrowlingson @Nowosad @Robinlovelace @gilbertocamara ?

**Comment by barryrowlingson** [comment]
I think I only suggested the opposite sense for NA and NULL to keep compatibility with what happens already with basemaps! But if there's other places they are used then keeping consistent meanings is probably better.

**Comment by mtennekes** [comment]
> The question: wouldn't this change impact many existing code bases?

Do you mean with the existing v4 code bases?

I've checked with v3:

```r
tm_shape(World) +
    tm_polygons()
```

![image](https://github.com/user-attachments/assets/85f51100-0399-465c-b54b-4fae4a1140bf)


```r
tm_shape(World) +
	tm_polygons(col = NA)
```
![image](https://github.com/user-attachments/assets/90b02ec2-e411-4a44-9c04-d74f982d823b)

```r
tm_shape(World) +
	tm_polygons(col = NULL)
#> Warning message: In rep(x, length.out = nx) : 'x' is NULL so the result will be NULL
```

![image](https://github.com/user-attachments/assets/8474f5d8-5c7e-4761-9b17-494eb4c6bd4c)




> For the context, I added examples of the current behavior. The question: wouldn't this change impact many existing code bases?
> 
> ```r
> library(tmap)
> data(World)
> 
> tm_shape(World) +
>         tm_polygons()
> ```
> 
> ![](https://camo.githubusercontent.com/35e866c598409de25d059f70481264578005dbb1948538008017420496ce1525/68747470733a2f2f692e696d6775722e636f6d2f7463676c6349742e706e67)
> 
> ```r
> tm_shape(World) +
>         tm_polygons(fill = NA)
> ```
> 
> ![](https://camo.githubusercontent.com/278e0bb3dd37746b4e9ebccebe7c2c66ea0da2353459a2c67ac6dc7db9b6d82c/68747470733a2f2f692e696d6775722e636f6d2f775a38416735632e706e67)
> 
> ```r
> tm_shape(World) +
>         tm_polygons(fill = NULL)
> #> Error in val[[1]]: subscript out of bounds
> ```

**Comment by Nowosad** [comment]
👍🏻 Ok, now I understand the need for the consistency here (I've been using tmap4 for so long that I do not remember the old ways of doing things)

**Comment by barryrowlingson** [comment]
As a clarification, when you say "tmap4" does that mean version 3.99.xxx as is the current master branch? Or is that the last ever 3.x version? There is a v4 branch with somewhat older changes, but the instructions for "Installation of tmap (version 4) is straightforward:" seem to get master from github.

**Comment by Robinlovelace** [comment]
> As a clarification, when you say "tmap4" does that mean version 3.99.xxx as is the current master branch? 

Yes that's correct. v4 is on the main branch.

**Comment by barryrowlingson** [comment]
I don't see a "main" branch, only a "master" branch and a "v4" branch... I'm assuming the "v4" branch is stale in some sense, that the "master" branch is the latest tmap version 4 code base, and that someone didn't change all their "master" branches to "main" :)

**Comment by Robinlovelace** [comment]
`master` is the main branch. The default one.

**Comment by Robinlovelace** [comment]
> I'm assuming the "v4" branch is stale

Correct. I just deleted it to avoid others getting confused, thanks for the nudge.

**Comment by Nowosad** [comment]
I've tried it, and there seems to be an issue for a non-global case:

```r
library(tmap)

# works fine
(tm = tm_shape(NLD_dist) +
    tm_polygons() +
    tm_basemap("OpenStreetMap"))

# works not fine
(tm2 = tm +
    tm_grid() +
    tm_basemap(NULL))
```

**Comment by mtennekes** [comment]
I get this for the second plot:

![image](https://github.com/user-attachments/assets/ff3d4cff-1743-4fe2-a756-4b989f2367f8)

Seems alright to me.

**Comment by Nowosad** [comment]
@mtennekes the issue is in the view mode:

```r
library(tmap)
tmap_mode("view")
# works fine
(tm = tm_shape(NLD_dist) +
    tm_polygons() +
    tm_basemap("OpenStreetMap"))

# works not fine (cannot see the polygons)
(tm2 = tm +
    tm_grid() +
    tm_basemap(NULL))
```

**Comment by mtennekes** [comment]
Funny. It seems to be an issue with `NLD_dist`.

`NLD_muni` and `NLD_prov` both work well. (Also for you?)

`tm_grid` is not supposed to be working with projected shapes in view mode, but tm_mouse_coordinates indicates that the shape is correctly projected:

```{r}
tm_shape(NLD_muni) +
	tm_polygons() +
	tm_basemap(NULL) + 
	tm_mouse_coordinates()
```

Glad this is not a major bug (bad timing:-))

**Comment by mtennekes** [comment]
> It seems to be an issue with `NLD_dist`.

Nope. `NLD_dist` is fine.

Apparently projected shapes in view mode don't work with WebGL enabled (which is the default for large shapes, and `NLD_dist` is considered large). This also works:

```{r}
tm_shape(NLD_muni) +
	tm_polygons() +
	tm_basemap(NULL) + 
	tm_mouse_coordinates()
```

So still a bug, bug minor, and probably something upstream @tim-salabim does leafgl work with projected shapes and L.CRS.Simple?

**Comment by tim-salabim** [comment]
Yeah, seems like it's not compatible with `L.CRS.Simple`... I have no idea why, but a guess is that it doesn't handle the maps `latLngToLayerPoint()` function properly. Maybe @trafficonese has any ideas/suggestions?

Here's a minimal reprex

```r
library(leaflet)
library(leafgl)
library (tmap)
library(sf)

data("NLD_dist")

leaflet(options = leafletOptions(
  crs = leafletCRS(crsClass = "L.CRS.Simple")
  , minZoom = -1000
)) |>
  addGlPolygons(data = st_cast(NLD_dist, "POLYGON"))
```

**Comment by mtennekes** [comment]
webGL already disabled for projected crs in commit [429cd5a](https://github.com/r-tmap/tmap/commit/429cd5a6f1d217c75c3137b4373b90b20cbf47e1)
